### ISSUE TITLE
JS version is suboptimal

### ISSUE THREAD

**Issue by mihai1voicescu** [emojione/emojione#546]
Multiple improvements can be made to speedup the JS code:
1. The regex used in transformation functions (such as `shortnameToAscii`) should not contain the tags, only `:code:` alternatives. The callback of the replace function is called each time it catches a tag and returns just that. If this is actually necessary(can't think of a good reason beside leaving the tags intact) the callback of `ns.regShortNames` should make use of the associated capture group. This will allow the following contraction:
```
 str = str.replace(ns.regShortNames, function(shortname) {
            if( (typeof shortname === 'undefined') || (shortname === '') || (!(shortname in ns.emojioneList)) ) {
                // if the shortname doesnt exist just return the entire matchhju
                return shortname;
            }
```
to
```
 str = str.replace(ns.regShortNames, function(match, shortname) {
           if (!shortname)
               return match;
```
2. The same contraction should be applied to every callback function.
3. Replace callback functions should be pulled outside the containing functions scope. Each time a function is called (let's say `ns.shortnameToAscii` for example) an anonymous function is created causing overhead. If this is pulled outside the main function the function will be created only once and reused. This will trade a bit of ram for more performance.
4. Stop using closure variables where they are not needed. This can keep the interpreter from optimizing the code and inline-ing the functions.
5. Variables are calculated but are never used(eg: [1](https://github.com/emojione/emojione/blob/master/lib/js/emojione.js#L214))
6. Variables are calculated before deciding if they are usefull(eg: [1](https://github.com/emojione/emojione/blob/master/lib/js/emojione.js#L293), [2](https://github.com/emojione/emojione/blob/master/lib/js/emojione.js#L263))
7. Cache the result of https://github.com/emojione/emojione/blob/master/lib/js/emojione.js#L184
8. Regex can be improved by collapsing the alternatives.
* Minimum(very safe and low complexity) pull the ':' from the alternatives: `/:a:|:b:/` -> `/:(?:a|b):/`
* Maximum(using a collapse script for alternatives) `/:aabc:|:aafg:/` -> `/:aa(?:bc|fg):/`

I could help with a pull request if you want 😃

**Comment by mihai1voicescu** [comment]
Just gonna do the PR 😄.

**Comment by mihai1voicescu** [comment]
Also I think there is a bug with keys that have + in them on master(`emojione.toImage(":+1_tone1:")`). The problem is in the searched alias array + is replaced. Will fix this too.

**Comment by anamauad** [comment]
@mihai1voicescu , I think you should not mix corrections for different issues, as `emojione.toImage(":+1_tone1:")` not being converted

**Comment by mihai1voicescu** [comment]
@anamauad you are right, but as I removed the linear search I see no point in providing the fix for it.
